### ISSUE TITLE
Integration of integrated objects

### ISSUE THREAD

**Issue by DHGK** [satijalab/seurat#1934]
Hi Seurat Community,

I wonder what would be the best practice for this scenario -- 

I have two biological replicates. Each replicate has data from 8 different timepoints. I have tried to integrate the data in the following ways:

1. Integrate 2*8 datasets together (16 -> 1)
2. First integrate each replicate separately, and then integrate 2 integrated replicates (16 -> 2 -> 1)
3. First integrate each timpoint separately, and then integrate 8 integrated timpoints (16 -> 8 -> 1)

Theoretically, would any one of the three ways be better than others? 

Really appreciate all your help!

**Comment by timoast** [comment]
Hi, there is likely very little difference between the above options in the final outcome. If for example you were interested in preserving the timepoint difference but eliminating the differences between replicates, you could integrate between replicates within a timepoint. If you're just interested in integrating everything together, any of the above options will work.

**Comment by liuxl18-hku** [comment]
Hi @timoast ,

Sorry for bothering u.

Actually, I need to integrate the integrated objects (generated by SelectIntegrationFeatures-> PrepSCTIntegration -> FindIntegrationAnchors("SCT", "rpca") -> IntegrateData), as I have too many cells (>300k) that the IntegratedData (actually Matrix package doesn't support the long vector as mentioned here #1029 ).

Now, I am doing the following:
1. Integerate data from each patients first (SelectIntegrationFeatures-> PrepSCTIntegration -> FindIntegrationAnchors("SCT", "rpca") -> IntegrateData)
2. Integrate the integrated data from 1. using same approach

In step 1., everything works perfectly OK (though the we used Seurat_3.0.2 under R343 for per-sample pre-process, but same as Seurat version of Integration), the step 2., however, I had some problems:

```r
sample.features <- SelectIntegrationFeatures(object.list = sample.list, nfeatures = 3000)
sample.list <- PrepSCTIntegration(object.list = sample.list, anchor.features = sample.features, verbose = TRUE)
sample.list <- lapply(X = sample.list, FUN = RunPCA, verbose = TRUE, features = sample.features)
sample.anchors <- FindIntegrationAnchors(object.list = sample.list, normalization.method = "LogNormalize", anchor.features = sample.features, verbose = TRUE, reduction = "rpca")
integrated <- IntegrateData(anchorset = sample.anchors, normalization.method = "SCT", verbose = TRUE) # Error in CellsByIdentities(object = object, cells = cells) : Cannot find cells provided
```

Session info:
```r
> sessionInfo()
R version 3.4.3 (2017-11-30)
Platform: x86_64-conda_cos6-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /home/liuxl18/anaconda3/envs/R343/lib/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Seurat_3.1.0.9004

loaded via a namespace (and not attached):
 [1] httr_1.4.0          tidyr_0.8.3         viridisLite_0.3.0  
 [4] jsonlite_1.6        splines_3.4.3       lsei_1.2-0         
 [7] R.utils_2.9.0       leiden_0.3.1        gtools_3.8.1       
[10] RcppParallel_4.4.3  Rdpack_0.11-0       assertthat_0.2.1   
[13] ggrepel_0.8.1       globals_0.12.4      pillar_1.4.2       
[16] lattice_0.20-35     glue_1.3.1          reticulate_1.13    
[19] digest_0.6.20       RColorBrewer_1.1-2  SDMTools_1.1-221.1 
[22] colorspace_1.4-1    cowplot_0.9.3       htmltools_0.3.6    
[25] Matrix_1.2-17       R.oo_1.22.0         plyr_1.8.4         
[28] pkgconfig_2.0.2     bibtex_0.4.2        tsne_0.1-3         
[31] listenv_0.7.0       purrr_0.3.2         scales_1.0.0       
[34] RANN_2.6.1          gdata_2.18.0        Rtsne_0.15         
[37] tibble_2.1.3        ggplot2_3.2.0       ROCR_1.0-7         
[40] pbapply_1.4-1       lazyeval_0.2.2      survival_2.41-3    
[43] magrittr_1.5        crayon_1.3.4        R.methodsS3_1.7.1  
[46] future_1.14.0       nlme_3.1-131        MASS_7.3-48        
[49] gplots_3.0.1.1      ica_1.0-2           tools_3.4.3        
[52] fitdistrplus_1.0-14 data.table_1.12.2   gbRd_0.4-11        
[55] stringr_1.4.0       plotly_4.9.0        munsell_0.5.0      
[58] cluster_2.0.6       irlba_2.3.3         compiler_3.4.3     
[61] rsvd_1.0.2          caTools_1.17.1.2    rlang_0.4.0        
[64] grid_3.4.3          ggridges_0.5.1      RcppAnnoy_0.0.12   
[67] htmlwidgets_1.3     igraph_1.2.4.1      bitops_1.0-6       
[70] npsurv_0.4-0        gtable_0.3.0        codetools_0.2-15   
[73] reshape2_1.4.3      R6_2.4.0            gridExtra_2.3      
[76] zoo_1.8-6           dplyr_0.8.3         uwot_0.1.3         
[79] future.apply_1.3.0  KernSmooth_2.23-15  metap_1.1          
[82] ape_5.3             stringi_1.4.3       parallel_3.4.3     
[85] Rcpp_1.0.2          sctransform_0.2.0   png_0.1-7          
[88] tidyselect_0.2.5    lmtest_0.9-37 
``` 

Any help is appreciated!

bless~

**Comment by liuxl18-hku** [comment]
@timoast ,

I solved the problem under the context of `Seurat_3.1.0.9004`, basically, I tried to update vst.set of `SCT` assay by myself:

```r
# code before is omitted
sample.anchors <- FindIntegrationAnchors(object.list = sample.list, normalization.method = "SCT", anchor.features = sample.features, verbose = T, reduction = "rpca")

# the part to update vst.set names
for(i in 1:length(sample.anchors@object.list)) {
    for(j in 1:length(sample.anchors@object.list[[i]]@assays$SCT@misc$vst.set)) {
        old_name <- rownames(sample.anchors@object.list[[i]]@assays$SCT@misc$vst.set[[j]]$cell_attr)
        rownames(sample.anchors@object.list[[i]]@assays$SCT@misc$vst.set[[j]]$cell_attr) <- paste0(old_name, "_", j, "_", i)
    }
}
integrated <- IntegrateData(anchorset = sample.anchors, normalization.method = "SCT", verbose = T)
```

bless~

**Comment by qdong2023** [comment]
Hi, @timoast, I'm integrating two batches of datasets, each of which contains several samples. The samples in the 1st batch have some variations, so I did the integration for this batch and got a good correction. However, when I did the integration for the integrated dataset and the 2nd batch, the result was the same as integrating the original two batches (without pre-integration). I am wondering if I should use the "RNA" assay or the "integrated" assay for the second round of integration (using the integrated datasets). Thanks!
